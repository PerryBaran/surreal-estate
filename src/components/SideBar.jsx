import React, { useRef } from "react";
import { Link, useLocation, useNavigate } from "react-router-dom";
import PropTypes from "prop-types";
import qs from "qs";
import style from "../styles/sideBar.module.css";
import { search as searchIcon } from "../media/icons";

const SideBar = ({
  options: { cities, types },
  userId,
  filterByFavourites,
  handleFilterFavourites,
}) => {
  const queryRef = useRef();
  const navigate = useNavigate();
  const { search } = useLocation();

  const buildQueryString = (operation, valueObj) => {
    const currentQueryParams = qs.parse(search, { ignoreQueryPrefix: true });

    const newQueryParams = {
      ...currentQueryParams,
      [operation]: JSON.stringify({
        ...JSON.parse(currentQueryParams[operation] || "{}"),
        ...valueObj,
      }),
    };

    return qs.stringify(newQueryParams, {
      addQueryPrefix: true,
      encode: false,
    });
  };

  const handleSearch = (e) => {
    e.preventDefault();
    const { value } = queryRef.current;

    const newQueryString = buildQueryString("query", {
      title: { $regex: value },
    });

    navigate(newQueryString);
  };

  return (
    <div className={style["side-bar"]}>
      {userId && (
        <label htmlFor="favourites" className={style["side-bar-favourites"]}>
          <h2 className={style["side-bar-heading"]}>Filter By Favourites</h2>
          <input
            type="checkbox"
            id="favourites"
            checked={filterByFavourites}
            onChange={handleFilterFavourites}
          />
        </label>
      )}
      <h2 className={style["side-bar-heading"]}>Filter By City</h2>
      <ul className={style["side-bar-list"]}>
        {cities.map((city) => {
          return (
            <li className={style["side-bar-item"]} key={city}>
              <Link
                to={buildQueryString("query", { city })}
                className={style["side-bar-link"]}
              >
                {city}
              </Link>
            </li>
          );
        })}
      </ul>
      <h2 className={style["side-bar-heading"]}>Filter By Type</h2>
      <ul className={style["side-bar-list"]}>
        {types.map((type) => {
          return (
            <li className={style["side-bar-item"]} key={type}>
              <Link
                to={buildQueryString("query", { type })}
                className={style["side-bar-link"]}
              >
                {type}
              </Link>
            </li>
          );
        })}
      </ul>
      <h2 className={style["side-bar-heading"]}>Sort By Price</h2>
      <ul className={style["side-bar-list"]}>
        <li className={style["side-bar-item"]}>
          <Link
            to={buildQueryString("sort", { price: -1 })}
            className={style["side-bar-link"]}
          >
            Ascending
          </Link>
        </li>
        <li className={style["side-bar-item"]}>
          <Link
            to={buildQueryString("sort", { price: 1 })}
            className={style["side-bar-link"]}
          >
            Descending
          </Link>
        </li>
      </ul>
      <h2 className={style["side-bar-heading"]}>Search by Title</h2>
      <form onSubmit={handleSearch} className={style["side-bar-search-form"]}>
        <div className={style["side-bar-search"]}>
          <label htmlFor="query">
            <input
              id="query"
              ref={queryRef}
              className={style["side-bar-search-input"]}
            />
          </label>
          <button type="submit" className={style["side-bar-search-submit"]}>
            <img alt="search" src={searchIcon} />
          </button>
        </div>
      </form>
    </div>
  );
};

SideBar.propTypes = {
  options: PropTypes.shape({
    cities: PropTypes.arrayOf(PropTypes.string),
    types: PropTypes.arrayOf(PropTypes.string),
  }).isRequired,
  filterByFavourites: PropTypes.bool.isRequired,
  handleFilterFavourites: PropTypes.func.isRequired,
  userId: PropTypes.string.isRequired,
};

export default SideBar;
