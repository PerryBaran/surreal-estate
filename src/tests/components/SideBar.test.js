import React from "react";
import { BrowserRouter as Router } from "react-router-dom";
import { fireEvent, render, screen } from "@testing-library/react";
import SideBar from "../../components/SideBar";

const RenderWithRouter = (props) => {
  return (
    <Router>
      <SideBar {...props} />
    </Router>
  );
};

describe("SideBar", () => {
  describe("falsey userId", () => {
    const validProps = {
      cities: ["Leeds", "Manchester"],
      types: ["Flat", "Detached"],
      userId: "",
      filterByFavourites: false,
      handleFilterFavourites: jest.fn(),
    };

    test("snapshot", () => {
      const { asFragment } = render(<RenderWithRouter {...validProps} />);

      expect(asFragment()).toMatchSnapshot();
    });

    test("renders correctly", () => {
      render(<RenderWithRouter {...validProps} />);

      expect(
        screen.queryByText(/fitler by favourites/i)
      ).not.toBeInTheDocument();
      expect(screen.getByText(/filter by city/i)).toBeInstanceOf(
        HTMLHeadingElement
      );
      expect(screen.getAllByRole("link")).toHaveLength(
        validProps.cities.length + validProps.types.length + 2 // Link for each city, type and price (ascending, descending)
      );
      validProps.cities.forEach((city) => {
        expect(screen.getByText(city)).toHaveAttribute(
          "href",
          `/?query={"city":"${city}"}`
        );
      });
      validProps.types.forEach((type) => {
        expect(screen.getByText(type)).toHaveAttribute(
          "href",
          `/?query={"type":"${type}"}`
        );
      });
      expect(screen.getByText(/sort by price/i)).toBeInstanceOf(
        HTMLHeadingElement
      );
      expect(screen.getByText(/ascending/i)).toHaveAttribute(
        "href",
        `/?sort={"price":-1}`
      );
      expect(screen.getByText(/descending/i)).toHaveAttribute(
        "href",
        `/?sort={"price":1}`
      );
      expect(screen.getByText(/search by title/i)).toBeInstanceOf(
        HTMLHeadingElement
      );
      expect(screen.getByRole("textbox")).toHaveAttribute("id", "query");
      expect(screen.getByText(/reset/i)).toBeInstanceOf(HTMLButtonElement);
      expect(screen.getByAltText("search")).toHaveAttribute(
        "src",
        "search.png"
      );
    });
  });

  describe("truthy userId", () => {
    const validProps = {
      cities: ["Leeds", "Manchester"],
      types: ["Flat", "Detached"],
      userId: "76yhh64b",
      filterByFavourites: false,
      handleFilterFavourites: jest.fn(),
    };

    test("snapshot", () => {
      const { asFragment } = render(<RenderWithRouter {...validProps} />);

      expect(asFragment()).toMatchSnapshot();
    });

    test("renders favourite checkbox", () => {
      render(<RenderWithRouter {...validProps} />);

      expect(screen.getByText(/filter by favourites/i)).toBeInTheDocument();
    });

    test("clicking links adds and removes from query string", () => {
      render(<RenderWithRouter {...validProps} />);

      const city = validProps.cities[0];
      const type = validProps.types[0];
      const cityLink = screen.getByText(city);
      const typeLink = screen.getByText(type);

      fireEvent.click(cityLink);
      expect(typeLink).toHaveAttribute(
        "href",
        `/?query={"city":"${city}","type":"${type}"}`
      );
      fireEvent.click(cityLink);
      expect(typeLink).toHaveAttribute("href", `/?query={"type":"${type}"}`);
    });

    test("reset button", () => {
      render(<RenderWithRouter {...validProps} />);

      const city = validProps.cities[0];
      const type = validProps.types[0];
      const cityLink = screen.getByText(city);
      const typeLink = screen.getByText(type);
      const reset = screen.getByText(/reset/i);

      fireEvent.click(cityLink);
      expect(typeLink).toHaveAttribute(
        "href",
        `/?query={"city":"${city}","type":"${type}"}`
      );
      fireEvent.click(reset);
      expect(typeLink).toHaveAttribute("href", `/?query={"type":"${type}"}`);
    });
  });
});
