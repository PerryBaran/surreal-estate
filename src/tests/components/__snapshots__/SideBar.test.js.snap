// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SideBar falsey userId snapshot 1`] = `
<DocumentFragment>
  <div
    class="side-bar"
  >
    <h2
      class="side-bar__heading"
    >
      Filter By City
    </h2>
    <ul>
      <li>
        <a
          class="side-bar__link null"
          href="/?query={\\"city\\":\\"Leeds\\"}"
        >
          Leeds
        </a>
      </li>
      <li>
        <a
          class="side-bar__link null"
          href="/?query={\\"city\\":\\"Manchester\\"}"
        >
          Manchester
        </a>
      </li>
    </ul>
    <h2
      class="side-bar__heading"
    >
      Filter By Type
    </h2>
    <ul>
      <li>
        <a
          class="side-bar__link null"
          href="/?query={\\"type\\":\\"Flat\\"}"
        >
          Flat
        </a>
      </li>
      <li>
        <a
          class="side-bar__link null"
          href="/?query={\\"type\\":\\"Detached\\"}"
        >
          Detached
        </a>
      </li>
    </ul>
    <h2
      class="side-bar__heading"
    >
      Sort By Price
    </h2>
    <ul>
      <li>
        <a
          class="side-bar__link null"
          href="/?sort={\\"price\\":-1}"
        >
          Ascending
        </a>
      </li>
      <li>
        <a
          class="side-bar__link null"
          href="/?sort={\\"price\\":1}"
        >
          Descending
        </a>
      </li>
    </ul>
    <h2
      class="side-bar__heading"
    >
      Search by Title
    </h2>
    <div
      class="side-bar__search"
    >
      <input
        class="side-bar__search__input"
        id="query"
      />
      <button
        class="side-bar__search__submit"
        type="submit"
      >
        <img
          alt="search"
          src="search.png"
        />
      </button>
    </div>
    <button
      class="side-bar__reset"
      type="button"
    >
      Reset
    </button>
  </div>
</DocumentFragment>
`;

exports[`SideBar truthy userId snapshot 1`] = `
<DocumentFragment>
  <div
    class="side-bar"
  >
    <label
      class="side-bar__favourites"
      for="favourites"
    >
      <h2
        class="side-bar__heading"
      >
        Filter By Favourites
      </h2>
      <input
        id="favourites"
        type="checkbox"
      />
    </label>
    <h2
      class="side-bar__heading"
    >
      Filter By City
    </h2>
    <ul>
      <li>
        <a
          class="side-bar__link null"
          href="/?query={\\"city\\":\\"Leeds\\"}"
        >
          Leeds
        </a>
      </li>
      <li>
        <a
          class="side-bar__link null"
          href="/?query={\\"city\\":\\"Manchester\\"}"
        >
          Manchester
        </a>
      </li>
    </ul>
    <h2
      class="side-bar__heading"
    >
      Filter By Type
    </h2>
    <ul>
      <li>
        <a
          class="side-bar__link null"
          href="/?query={\\"type\\":\\"Flat\\"}"
        >
          Flat
        </a>
      </li>
      <li>
        <a
          class="side-bar__link null"
          href="/?query={\\"type\\":\\"Detached\\"}"
        >
          Detached
        </a>
      </li>
    </ul>
    <h2
      class="side-bar__heading"
    >
      Sort By Price
    </h2>
    <ul>
      <li>
        <a
          class="side-bar__link null"
          href="/?sort={\\"price\\":-1}"
        >
          Ascending
        </a>
      </li>
      <li>
        <a
          class="side-bar__link null"
          href="/?sort={\\"price\\":1}"
        >
          Descending
        </a>
      </li>
    </ul>
    <h2
      class="side-bar__heading"
    >
      Search by Title
    </h2>
    <div
      class="side-bar__search"
    >
      <input
        class="side-bar__search__input"
        id="query"
      />
      <button
        class="side-bar__search__submit"
        type="submit"
      >
        <img
          alt="search"
          src="search.png"
        />
      </button>
    </div>
    <button
      class="side-bar__reset"
      type="button"
    >
      Reset
    </button>
  </div>
</DocumentFragment>
`;
